Feature/killswitch

NickGreen opened the pull request:
## Synopsis

Simplified version of the centralized settings here: https://github.com/a8cteam51/plugin-autoupdate-filter/pull/24

The point of this is to get an MVP up and running, with the most important feature, and then we can discuss and add the "Swiss Army Knife" features later.

## To Test

- This plugin is activated on opsoasis: https://github.com/a8cteam51/team51-autoupdate-settings
- The settings page is here, with a toggle switch: https://opsoasis.wpspecialprojects.com/wp-admin/tools.php?page=team51_autoupdate_settings_options
- The API endpoint is here: https://opsoasis.wpspecialprojects.com/wp-json/custom/v1/get_autoupdate_settings/

When that toggle switch is toggled on, then all autoupdates should stop happening, and a notice should appear on the top of the plugins page.

## Details

- **Caching:** The settings are cached, causing the killswitch to not immediately take effect. This is likely because the settings are retrieved when the object is created. However, I also tried it by retrieving the settings every single time the update filters ran, and it still seemed to not take effect immediately. I can deactivate/reactivate the Plugin Autoupdate Filter plugin to get it to register changes to the settings. Any thoughts on how to do this better are encouraged.
- **Naming of the killswitch:** as is being discussed in the other PR, I'm still waffling over the naming of the killswitch being confusing with the boolean return for the autoupdate filter itself. (i.e. Killswitch is "on" and return is "false")

Any other feedback is helpful, with the goal of getting out a minimum viable product that includes the killswitch, and we can add more features later :)

NickGreen committed:
Get JSON from centralized location

NickGreen committed:
Update class-plugin-autoupdate-filter.php

NickGreen commented:
Example of a similar UI feature in another plugin
![SCR-20240301-ngat](https://github.com/a8cteam51/plugin-autoupdate-filter/assets/2067992/16f9a517-3a4e-433e-9613-e1fe71742eb9)

ahegyes commented:
> The settings are cached, causing the killswitch to not immediately take effect. 

Where are they cached exactly? Do you mean the nginx cache for `https://opsoasis.wpspecialprojects.com/wp-json/custom/v1/get_autoupdate_settings/`? For that we'd need to run `wp_cache_flush` on OpsOasis whenever a setting changes. Or maybe there is a more specific function provided by `batcache` to only flush that one URL...

However, I _think_ we should also include a transient cache after retrieving the setting. Even if just a 5 minute one ... to prevent a high-traffic site potentially bringing down OpsOasis if something goes wrong with the caching ...

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -37,6 +39,57 @@ public function init() {
+	 */
+	public function auto_update_killswitch( $update, $item ) {
+		if ( ! isset ( $this->settings['team51_autoupdate_settings_disable_all_toggle'] ) ) {
+			return ;
```
Since this function is added to a filter, we should return something here or we risk breaking something (at least PHP warnings would show up); in this case, the natural option would be `$update`;

But given how this function is constructed, we can probably give up this `if` check altogether ...

NickGreen replied:
Good catch! Removed redundant code.

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -199,6 +248,27 @@ function() use ( $slug ) {
+			add_action(
+				'admin_notices',
+				function() {
+					echo '<div class="error"><p><strong style="color:red;"> Caution:</strong> All autoupdates are currently deactivated. Please contact the WordPress Special Projects team before manually updating.</p></div>';
```
Shouldn't the classes on this notice be `notice notice-error`? https://developer.wordpress.org/reference/hooks/admin_notices/

NickGreen replied:
Yes! Fixed.

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -37,6 +39,57 @@ public function init() {
+	/**
+	 * Load settings from the centralized settings page
+	 */
+	private function get_auto_update_settings() {
```
This function has a return value only when there's an error. In other words, not all execution paths return something. Moreover, there is nothing being done with the return value. 

I suggest that the function either returns `array|WP_Error` or we throw an exception instead of returning some strings. In both cases, the error handling would happen in the calling function.

NickGreen replied:
Thanks for catching this, as it's clearly not perfectly thought out.

Re: "return;"  My understanding is that using "return" is the accepted way to exit out of a function early, and returning a value is something that's optional. Always open to learning otherwise, though :) 

I think this is closer to what I was trying to do: set the "settings" to have an error value if there's an error. such as:

```PHP
private function get_auto_update_settings() {
  $endpoint_url = 'https://opsoasis.wpspecialprojects.com/wp-json/custom/v1/get_autoupdate_settings/';
  $response     = wp_remote_get( $endpoint_url );
  
  // Check for `WP_Error`
  if ( is_wp_error( $response ) ) {
	  $this->settings = [
		  'error' => $response->get_error_message(),
	  ];
	  return;
  }
  
  // Check for non-200 status code
  $response_code = wp_remote_retrieve_response_code($response);
  if ( 200 !== $response_code ) {
	  $this->settings = [
		  'error' => "HTTP request returned status code {$response_code}.",
	  ];
	  return;
  }
  
  // Check if response body exists and is not empty
  $body = wp_remote_retrieve_body($response);
  if ( empty( $body ) ) {
	  $this->settings = [
		  'error' => 'API returned an empty response body.',
	  ];
	  return;
  }
  
  // Try to decode JSON and check for errors
  $data = json_decode( $body, true );
  if ( json_last_error() !== JSON_ERROR_NONE ) {
	  $this->settings = [
		  'error' => 'Failed to decode JSON: ' . json_last_error_msg(),
	  ];
	  return;
  }
  
  $this->settings = $data;
}
 ```

This way, we can check for an 'error' value in `$this->settings` later, in order to pause all autoupdates if there's been an error getting the settings, such as:

```PHP
public function auto_update_killswitch( $update, $item ) {

  if ( $this->settings['error'] || 'on' === $this->settings['team51_autoupdate_settings_disable_all_toggle'] ) {
	  return false;
  }
  
  return $update;
}
```

And then also add it to an admin notice, like:

```PHP
public function autoupdate_settings_admin_notice() {
  $message = '';
  if ( $this->settings['error'] ) {
	  $message = 'Error retrieving autoupdate settings (' . $this->settings['error'] . '). ' ;
  }
 ...
```

Thoughts?

ahegyes replied:
> My understanding is that using "return" is the accepted way to exit out of a function early, and returning a value is something that's optional.

Well yes, but also no 😅 Technically, [PHP functions **always** return a value](https://www.php.net/manual/en/functions.returning-values.php). If you use `return;` or simply never return at all, then the return value will be `null`.

Once PHP7 came around, it became possible to type-hint `void` as the function's return value and that would throw an error if your function contained anything other than `return;`. However, [even in this instance](https://stackoverflow.com/a/40960562), you could catch the function's output and it would be `null`.

So the return value is definitely optional since you always have a default of `null` regardless.

-----

That being said, it's typically bad practice to have paths in your code that explicitly return something and others that don't. If you use exclusively `return;` in the code, that's fine (the function's return value is then `void` basically). If you never use `return` at all, that's fine. But when you mix `return;` with `return x;` you start having to do more error-handling than it's worth. 

For example, in your original code, you would have two cases: a) the function returns a string which we should know to interpret as an error, or b) the function returns nothing which implies that it returns `null`. In **both cases**, the programmer picking up the code needs to know something: a) a string implies an error, b) if no string, then null because that's how PHP works (for now), and that means success. 

[PHPMD](https://phpmd.org/) would flag this with an error message like "Not all execution paths return a value". It's fine that none do, or that all do, but mixing them gets iffy. 

-----

I like your current code -- it fixes the mixed paths issue, but I think it's needlessly complicated. The `settings` object now fulfils two separate functions, can have two separate sets of properties, and requires checking which object we're currently working with at every step of the way. 

What about something like this:

```
private function get_auto_update_settings(): stdClass|WP_Error {
    $response = wp_remote_get(
        'https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/'
        array( 'Accept' => 'application/json' )
    );
    if ( is_wp_error( $response ) ) {
        return $response;
    }

    $response_code = wp_remote_retrieve_response_code( $response );
    $response_body = wp_remote_retrieve_body( $response );
 
    // Check that the response code is a 2xx code.
    if ( ! \str_starts_with( (string) $response_code, '2' ) ) {
        $response_message = wp_remote_retrieve_response_message( $response );
        return new WP_Error( $response_code, $response_message, $response_body );
    }

    // Parse the response body as JSON.
    try {
        $decoded_body = json_decode( $response_body, false, 512, JSON_THROW_ON_ERROR );
    } catch ( JsonException $exception ) {
        return new WP_Error( $exception->getCode(), $exception->getMessage(), $exception );
    }

    return $decoded_body;
}

public function init(): void {
    $this->settings = $this->get_auto_update_settings();
    if ( is_wp_error( $this->settings ) ) {
        // Optionally error log, output as an admin notice, no whatever. 
        error_log( "❌ Plugin Autoupdate Filter API Error: { $this->settings->get_error_code() } { $this->settings->get_error_message() }" ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
        add_action(
            'admin_notices',
            function() {
	        $message = $this->settings->get_error_message();
	        printf( '<div class="notice notice-error"><p>%1$s</p></div>', esc_html( $message ) );
            }
        );

        // Disable all updates.
        $this->settings = (object) array( 'disable_all' => true );
    }
  
    // Continue with plugin initialization.
}
```

That would be the "truly" WordPress way of doing things before PHP even supported exceptions. If you want to use a more modern approach similar to how WooCommerce handles things, then may I propose:

```
private function get_auto_update_settings(): stdClass {
    $response = wp_remote_get(
        'https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/'
        array( 'Accept' => 'application/json' )
    );
    if ( is_wp_error( $response ) ) {
        throw new RuntimeException( $response->get_error_message(), $response->get_error_code() );
    }

    $response_code = wp_remote_retrieve_response_code( $response );
    $response_body = wp_remote_retrieve_body( $response );
 
    // Check that the response code is a 2xx code.
    if ( ! \str_starts_with( (string) $response_code, '2' ) ) {
        $response_message = wp_remote_retrieve_response_message( $response );
        throw new Exception( $response_message, $response_code );
    }

    // Parse the response body as JSON.
    return json_decode( $response_body, false, 512, JSON_THROW_ON_ERROR );
}

public function init(): void {
    try {
        $this->settings = $this->get_auto_update_settings();
    } catch ( Exception $exception ) {
        // Optionally error log, output as an admin notice, no whatever. 

        // Disable all updates.
        $this->settings = (object) array( 'disable_all' => true );
    }
   
     // Continue with plugin initialization.
}
```

Either of these solutions would simplify your code significantly because:

1. All the error checking and logging would happen in one single place.
2. You can start using the settings property **without** doing error checking throughout your code.

You'll notice that I used a `stdClass` object instead of an array. That was just to keep everything an object but ultimately it doesn't matter. An array might be easier to use since you could `??` any unset values to a default inline.

NickGreen replied:
This is great! I've reworked the code to follow this paradigm

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -37,6 +39,57 @@ public function init() {
+	 * Load settings from the centralized settings page
+	 */
+	private function get_auto_update_settings() {
+		$endpoint_url = 'https://opsoasis.wpspecialprojects.com/wp-json/custom/v1/get_autoupdate_settings/';
```
What do you think of renaming the endpoint: `/wp-json/wpcomsp/v1/autoupdate-settings/`? It's more inline with the REST endpoints defined by the `OpsOasis Endpoints` plugin but it shouldn't matter ultimately.

ahegyes replied:
Just thought of `wpcomsp/auto-update-plugin/v1/settings` 😄 (or `autoupdate-plugin`?)

The other endpoints are like `wpcomsp/<service>/v1/<resource>`, e.g. `wpcomsp/pressable/v1/sites`, so this would fit in perfectly. Of course, doesn't matter much, just thought I'd bring it up since this would be tricky to change after the first deployment.

NickGreen replied:
Valid! Updated endpoint to be `https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/`

ahegyes commented:
The setting name is extremely verbose 😅 I don't think we need the `team51_autoupdate_settings_` prefix for every setting given that it would apply to **all the settings** and we're retrieving them from a special endpoint already mentioning this. Ultimately it doesn't matter but these very long names are likely to lead to bug introduced by typos 😕

NickGreen commented:
@ahegyes 

> Where are they cached exactly?

The problem isn't on the opsoasis side of things; the wp-json endpoint is updating correctly and quickly each time. 

The lag is happening on the site where the Plugin Autoupdate Filter code is running. I _think_ what is happening is that when an instance of `class Plugin_Autoupdate_Filter` is initialized, that object persists for a while. Since the settings are pulled in when the object is created, then whatever settings were current at the time the object was created are used. Does that make sense?

NickGreen committed:
Updated error handling and function naming

NickGreen committed:
New endpoint and fixed admin notice

NickGreen committed:
Fixed new endpoint, removed setting prefix

NickGreen commented:
@ahegyes 

## Updates

1. New endpoint is here: [https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/](https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/)
2. Error handling discussion in https://github.com/a8cteam51/plugin-autoupdate-filter/pull/26#discussion_r1511854817
3. Caching discussion in https://github.com/a8cteam51/plugin-autoupdate-filter/pull/26#issuecomment-1977424853

ahegyes commented:
> The lag is happening on the site where the Plugin Autoupdate Filter code is running. I think what is happening is that when an instance of class Plugin_Autoupdate_Filter is initialized, that object persists for a while. Since the settings are pulled in when the object is created, then whatever settings were current at the time the object was created are used. Does that make sense?

Hmmm ... unfortunately, it does **not** make sense 😢 Unless cached in the object cache, PHP objects don't persist past their current request unlike objects running on a Java Tomcat or ASP.NET server.

I also checked the code of `wp_remote_get` to make sure that it doesn't do any caching and I couldn't find anything 😕 

I haven't tested this to confirm but if you think the plugin is somehow using "old" settings even though OpsOasis has updated, then we need to investigate exactly why 😢 I still think some batcache feature would be the culprit...

I still think we should incorporate a ~5 minute transient ourselves for the settings so we only get a maximum of ~1000 requests every 5 minutes to OpsOasis (roughly the number of sites we run the plugin on) instead of any number of requests proportional to the number of visitors that our managed sites get, but that's a separate conversation 😄

NickGreen committed:
Reworked to handle errors

NickGreen commented:
@ahegyes 

1. I've reworked the code to match your error handling paradigm
2. I figured out the caching problem - it was on opsoasis! I fixed it by clearing the object cache when that specific option is saved :) 

The latest version of the settings plugin is on opsoasis and this branch of Plugin Autoupdate Filter can be loaded on any site to test.

NickGreen committed:
Error handling

ahegyes commented:
@NickGreen 

> I figured out the caching problem - it was on opsoasis! I fixed it by clearing the object cache when that specific option is saved :)

That makes a lot of sense! Especially since my `curl` tests were showing that the endpoint was being cached ... glad this was figured out 😄 

> I've reworked the code to match your error handling paradigm

I'm still not a fan of the fact that error handling happens in two places (lines 85 and 26 when _just_ 26 would be enough) nor of the fact that `settings` can still have two separate forms, **but** I agree that this is much better and that it works. Happy to approve it as-is.

Big fan of the renaming of the class methods though!

NickGreen committed:
Further cleanup of exception handling

NickGreen commented:
@ahegyes Thank you for your persistence, sorry to make you look through this so many times. I'm definitely slowly understanding your suggestions fully, without just copy/pasting them.

I've updated the code, and I'm hoping you'll have time to look at it again. No hurry, BTW :)

The only odd thing that I wasn't able to understand fully was how, when the JSON is empty, line 88 would not return an object, so starting on line 91, I had to set it as an object so that the function would return an object. Maybe that's not a big deal; it certainly works as-is.

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,104 @@ class Plugin_Autoupdate_Filter {
+	 */
+	private function get_auto_update_settings(): stdClass|WP_Error {
+
+		$response = wp_remote_get(
```
Since we're using a static string, `wp_remote_get` is fine ... but I want to propose that we switch to `wp_safe_remote_get` ([source](https://developer.wordpress.org/reference/functions/wp_safe_remote_get/)) simply because it would be better for the future if anything changes 😄

NickGreen replied:
Done!

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,104 @@ class Plugin_Autoupdate_Filter {
+
+		$decoded_body = json_decode( $response_body, false, 512, JSON_THROW_ON_ERROR );
+
+		// if the settings are empty, we still need to return an object
```
If `$response_body` is empty, then `json_decode` will throw an error and it will be caught. You can try it out yourself here: https://onlinephp.io/c/df536

Then, you set the function return type to be `stdClass|WP_Error`. If the JSON deserializes to something else, like an array, PHP itself will throw an exception! You can try that here: https://onlinephp.io/c/88fcd

So basically, you can immediately return line 88. If `$response_body` contains anything other than a JSON object, line 22 will catch that. Easy error handling! 😄

NickGreen replied:
I've tried this several times, and if I don't cast it specifically as an object, this is the error I get:

`PHP Fatal error:  Uncaught TypeError: Plugin_Autoupdate_Filter::get_auto_update_settings(): Return value must be of type stdClass, string returned in /Users/nickgreen/Local Sites/testi/app/public/wp-content/plugins/plugin-autoupdate-filter/class-plugin-autoupdate-filter.php:101`

To reproduce:
1. Toggle off the "disable all autoupdates" toggle
2. See that the JSON being returned from the site is `""`
3. Get that return value error, because that `json_decode` line is returning a string 🤔

ahegyes replied:
The fix is in line 22. Change `catch ( Exception $exception )` to `catch ( Exception|Error $exception )`.

**Or** simply change it to `catch ( Throwable $e )` since both the [`Exception`](https://www.php.net/manual/en/class.exception.php) and [`Error`](https://www.php.net/manual/en/class.error.php) classes in PHP implement [the `Throwable` interface](https://www.php.net/manual/en/class.throwable.php).

ahegyes replied:
Based on the docs, an `Error` is the base class for all internal PHP errors(like the returned type being wrong), whereas `Exception` is the base class for all user errors.

NickGreen replied:
The problem here is that it shouldn't be an error at all. 

When I toggle the switch off in the settings (meaning no settings are actually set), then it returns an empty string there. But this isn't an indication of an error, and I don't want the function to treat it as an error. I want it to interpret it as "No error, but no settings are set". 

Anyway, we can fix this in a future iteration, and I'm interested in getting something out the door soon, so I'm going to stick a pin in it for later. :)

ahegyes replied:
> and I don't want the function to treat it as an error

I agree completely! 

> then it returns an empty string there

I would consider this a bug on OpsOasis' end though. If the switch is turned off, you'd still expect a valid object as a return value. But the setting is set to "false" instead of "true".

NickGreen replied:
Good point, maybe I'm looking at this from the wrong end entirely, and should be making sure that the data returned from opsoasis is correct.

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,104 @@ class Plugin_Autoupdate_Filter {
+	/**
+	 * Load settings from the centralized settings page
+	 */
+	private function get_auto_update_settings(): stdClass|WP_Error {
```
What do you think about adding a 5 minute [transient](https://developer.wordpress.org/apis/transients/) to this function? I want to avoid a case where a few high-traffic sites generate so much traffic to OpsOasis that everything comes to a halt...

NickGreen replied:
Done!

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,104 @@ class Plugin_Autoupdate_Filter {
+	/**
+	 * Load settings from the centralized settings page
+	 */
+	private function get_auto_update_settings(): stdClass|WP_Error {
```
You can change the return type of the function to `stdClass` now since it never returns `WP_Error` ... it throws exceptions instead (which is probably better anyway) 😄

NickGreen replied:
Good catch!

ahegyes commented:
@NickGreen Very big fan of the current iteration! 😄 I left a few comments but 👍 from me.

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,104 @@ class Plugin_Autoupdate_Filter {
 
 		// Replace automatic update wording on plugin management page in admin
-		add_filter( 'plugin_auto_update_setting_html', array( $this, 'plugin_autoupdate_filter_custom_setting_html' ), 11, 3 );
+		add_filter( 'plugin_auto_update_setting_html', array( $this, 'custom_setting_html' ), 11, 3 );
```
Since we're renaming methods ... I already like the new names, but wanted to bring up my naming scheme since it seems the appropriate time. When it comes to naming filters, I usually name it something like `filter_<value>` because the name of the method will then tell me that this is a) a filter not an action, b) what/why are we filtering.

This particular one, for example, I would name `filter_plugin_auto_update_settings_html`. The ones below I'd name `filter_auto_update_emails` and then `filter_auto_updates_debug_email`. 

The next action on `admin_init` would be something like `output_auto_updates_disabled_notice`. When looking at the function name, I'd immediately know that a) it's an action, not a filter, b) it outputs something to the screen, c) that something is most likely an admin notice;

Just a suggestion that I thought was timely, no need to action on it. Love the current changes already!

NickGreen replied:
Suggestions taken!

NickGreen committed:
Added 5 minute transient

NickGreen committed:
Version bump

NickGreen committed:
Update README.md

NickGreen committed:
Account for empty string

NickGreen committed:
Save object in transient

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,115 @@ class Plugin_Autoupdate_Filter {
+	private function get_auto_update_settings(): stdClass {
+
+		// Try getting the settings from the transient first
+		$transient_key = 'auto_update_settings';
```
Could we add a prefix to this key name? Just in case 😅 I'm good with either `team51_` **or** `wpcomsp_` (I prefer the later but it doesn't matter).

NickGreen replied:
Good idea! Changed to `wpcomsp_auto_update_settings`

ahegyes commented on class-plugin-autoupdate-filter.php:
```diff
@@ -14,29 +14,115 @@ class Plugin_Autoupdate_Filter {
+		$transient_key = 'auto_update_settings';
+		$settings      = get_transient( $transient_key );
+
+		if ( false === $settings || "" === $settings ) {
```
Why check for these two conditions specifically? Couldn't we just do `if ( empty( $settings ) )`?

NickGreen replied:
Yes, good catch! Updated.

NickGreen committed:
Small cleanup

NickGreen commented:
OK, I've tested this locally to ensure that autoupdates still do work as intended. 

I've also loaded version 1.5.0 (the version with the killswitch functionality) on a single production site before we roll it out to all of them: https://americanperimetertrail.org/wp-admin/plugins.php

I'll check back in this week to ensure that it's not causing any problems, then will probably merge this to trunk, which will cause all of the existing copies of this plugin to autpopdate themselves.

**This also means that we can no longer mess with the opsoasis killswitch without it affecting prod sites.**

ahegyes commented:
> This also means that we can no longer mess with the opsoasis killswitch without it affecting prod sites.

The worst that can happen is that auto-updates will **stop**, right? 😅 Because the plugin won't be able to pull in the settings and will default to ... pausing, no?

NickGreen commented:
Ha, that's true. I'll roll this out today 👍
